mms/predict_frontend.py

```diff
@@ -0,0 +1,165 @@
+                                                        '%s is expected but %s is given.' % \
+                                                        (name, input_type, mime_type)
+                        file_data = input_file.read()
+                        assert isinstance(file_data, (str, bytes)), 'Image file buffer should be type str or ' \
+                                                                    'bytes, but got %s' % (type(file_data))
+                    else:
+                        form_data = self.handler.get_form_data(name)
+                        if form_data:
+                            # pylint: disable=deprecated-method
+                            file_data = base64.decodestring(self.handler.get_form_data(name))
+                        else:
+                            raise ValueError('This end point is expecting a data_name of %s. '
+                                             'End point details can be found here:http://<host>:<port>/api-description'
+                                             % name)
+                    input_data.append(file_data)
+            except Exception as e:  # pylint: disable=broad-except
+                if model_name + '_Prediction4XX' in MetricsManager.metrics:
+                    MetricsManager.metrics[model_name + '_Prediction4XX'].update(metric=1)
+                logger.error(str(e))
+                abort(400, str(e))
+        else:
+            msg = '%s is not supported for input content-type' % input_type
+            if model_name + '_Prediction5XX' in MetricsManager.metrics:
+                MetricsManager.metrics[model_name + '_Prediction5XX'].update(metric=1)
+            logger.error(msg)
+            abort(500, "Service setting error. %s" % (msg))
+
+        if self.batching:
+            response = self._batch_predict(input_data, input_type, output_type, model_name)
+        else:
+            response = self._single_predict(input_data, model_service, model_name)
+
+        # Construct response according to output type
+        if output_type == 'application/json':
+            logger.info('Response is text.')
+        elif output_type == 'image/jpeg':
+            logger.info('Response is jpeg image encoded in base64 string.')
+        else:
+            msg = '%s is not supported for input content-type.' % output_type
+            if model_name + '_Prediction5XX' in MetricsManager.metrics:
+                MetricsManager.metrics[model_name + '_Prediction5XX'].update(metric=1)
+            logger.error(msg)
+            abort(500, "Service setting error. %s" % msg)
+
+        logger.debug("Prediction request handling time is: %s ms",
+                     ((time.time() - handler_start_time) * 1000))
+        return self.handler.jsonify({'prediction': response})
+
+    def _batch_predict(self, input_data, input_type, output_type, model_name):
+        _id = str(uuid.uuid4())
```

Comment:
Does this ensure uniqueness? The responses from model-services are mapped to a particular frontend thread based on this right?

Reply:
Yes, here is a SO answer on uuid4: https://stackoverflow.com/questions/703035/when-are-you-truly-forced-to-use-uuid-as-part-of-the-design/786541#786541

"Frankly, in a single application space without malicious actors, the extinction of all life on earth will occur long before you have a collision, even on a version 4 UUID, even if you're generating quite a few UUIDs per second."